ssh非默认端口不支持

sealos为2.0.4版本
sealos init --kubeadm-config kubeadm-config.yaml.tmpl \
    --master 192.168.52.51:60022 \
    --master 192.168.52.52:60022 \
    --master 192.168.52.53:60022 \
    --node 192.168.52.54:60022 \
    --user root \
    --passwd 1 \
    --version v1.14.4 \
    --pkg-url /root/kube1.14.4.tar.gz 

![image](https://user-images.githubusercontent.com/46241382/64905266-3c20bf00-d708-11e9-842d-5d447083492c.png)

2019-09-14 00:47:49 [INFO] [github.com/fanux/sealos/install/utils.go:80] 192.168.52.51:60022      kubeadm init --config=/root/kubeadm-config.yaml --experimental-upload-certs
2019-09-14 00:47:49 [DEBG] [github.com/fanux/sealos/install/utils.go:89] command result is: Flag --experimental-upload-certs has been deprecated, use --upload-certs instead
[apiServer.certSANs: Invalid value: "192.168.52.51:60022": altname is not a valid IP address, DNS label or a DNS label with subdomain wildcards: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'); a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\*\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), apiServer.certSANs: Invalid value: "192.168.52.52:60022": altname is not a valid IP address, DNS label or a DNS label with subdomain wildcards: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'); a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\*\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), apiServer.certSANs: Invalid value: "192.168.52.53:60022": altname is not a valid IP address, DNS label or a DNS label with subdomain wildcards: a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'); a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\*\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]

2019-09-14 00:47:49 [EROR] [github.com/fanux/sealos/install/utils.go:91]        Error exec command failed Process exited with status 3
panic: 1

goroutine 1 [running]:
github.com/fanux/sealos/install.Cmd(0xc000026400, 0x13, 0x926a6b, 0x4b, 0x0, 0x0, 0x0)
        /go/src/github.com/fanux/sealos/install/utils.go:92 +0x3b9
github.com/fanux/sealos/install.(*SealosInstaller).InstallMaster0(0xc0001184d0)
        /go/src/github.com/fanux/sealos/install/sealos.go:64 +0x113
github.com/fanux/sealos/cmd.glob..func3(0xcc9a00, 0xc0000a5560, 0x0, 0x12)
        /go/src/github.com/fanux/sealos/cmd/init.go:42 +0x17e
github.com/fanux/sealos/vendor/github.com/spf13/cobra.(*Command).execute(0xcc9a00, 0xc0000a5440, 0x12, 0x12, 0xcc9a00, 0xc0000a5440)
        /go/src/github.com/fanux/sealos/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/fanux/sealos/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xcc9c60, 0x6, 0x90dbf0, 0x1)
        /go/src/github.com/fanux/sealos/vendor/github.com/spf13/cobra/command.go:850 +0x2f8
github.com/fanux/sealos/vendor/github.com/spf13/cobra.(*Command).Execute(0xcc9c60, 0x407250, 0xc000074058)
        /go/src/github.com/fanux/sealos/vendor/github.com/spf13/cobra/command.go:800 +0x2b
github.com/fanux/sealos/cmd.Execute()
        /go/src/github.com/fanux/sealos/cmd/root.go:46 +0x2d
main.main()
        /go/src/github.com/fanux/sealos/main.go:20 +0x20

--- a/install/init.go
+++ b/install/init.go
@@ -36,9 +36,13 @@ func BuildInit(masters []string, nodes []string, vip, pkgUrl string) {
 
 //KubeadmConfigInstall is
 func (s *SealosInstaller) KubeadmConfigInstall() {
+	var masters []string
+	for _, h := range s.Masters {
+		masters = append(masters, IpFormat(h))
+	}
 	var templateData string
 	if KubeadmFile == "" {
-		templateData = string(Template(s.Masters, s.VIP, Version))
+		templateData = string(Template(masters, s.VIP, Version))
 	} else {
 		fileData, err := ioutil.ReadFile(KubeadmFile)
 		defer func() {
@@ -49,7 +53,7 @@ func (s *SealosInstaller) KubeadmConfigInstall() {
 		if err != nil {
 			panic(1)
 		}
-		templateData = string(TemplateFromTemplateContent(s.Masters, s.VIP, Version, string(fileData)))
+		templateData = string(TemplateFromTemplateContent(masters, s.VIP, Version, string(fileData)))
 	}
 	cmd := "echo \"" + templateData + "\" > /root/kubeadm-config.yaml"
 	Cmd(s.Masters[0], cmd)
@@ -57,7 +61,7 @@ func (s *SealosInstaller) KubeadmConfigInstall() {
 
 //InstallMaster0 is
 func (s *SealosInstaller) InstallMaster0() {
-	cmd := fmt.Sprintf("echo %s apiserver.cluster.local >> /etc/hosts", s.Masters[0])
+	cmd := fmt.Sprintf("echo %s apiserver.cluster.local >> /etc/hosts", IpFormat(s.Masters[0]))
 	Cmd(s.Masters[0], cmd)
 
 	cmd = `kubeadm init --config=/root/kubeadm-config.yaml --experimental-upload-certs`

--- a/install/join.go
+++ b/install/join.go
@@ -29,13 +29,13 @@ func (s *SealosInstaller) GeneratorToken() {
 
 //JoinMasters is
 func (s *SealosInstaller) JoinMasters() {
-	cmd := fmt.Sprintf("kubeadm join %s:6443 --token %s --discovery-token-ca-cert-hash %s --experimental-control-plane --certificate-key %s", s.Masters[0], JoinToken, TokenCaCertHash, CertificateKey)
+	cmd := fmt.Sprintf("kubeadm join %s:6443 --token %s --discovery-token-ca-cert-hash %s --experimental-control-plane --certificate-key %s", IpFormat(s.Masters[0]), JoinToken, TokenCaCertHash, CertificateKey)
 
 	for _, master := range s.Masters[1:] {
-		cmdHosts := fmt.Sprintf("echo %s apiserver.cluster.local >> /etc/hosts", s.Masters[0])
+		cmdHosts := fmt.Sprintf("echo %s apiserver.cluster.local >> /etc/hosts", IpFormat(s.Masters[0]))
 		Cmd(master, cmdHosts)
 		Cmd(master, cmd)
-		cmdHosts = fmt.Sprintf(`sed "s/%s/%s/g" -i /etc/hosts`, s.Masters[0], master)
+		cmdHosts = fmt.Sprintf(`sed "s/%s/%s/g" -i /etc/hosts`, IpFormat(s.Masters[0]), IpFormat(master))
 		Cmd(master, cmdHosts)
 	}
 }
@@ -45,7 +45,7 @@ func (s *SealosInstaller) JoinNodes() {
 	var masters string
 	var wg sync.WaitGroup
 	for _, master := range s.Masters {
-		masters += fmt.Sprintf(" --master %s:6443", master)
+		masters += fmt.Sprintf(" --master %s:6443", IpFormat(master))
 	}
 
 	for _, node := range s.Nodes {

--- a/install/sealos.go
+++ b/install/sealos.go
@@ -53,7 +53,7 @@ func (s *SealosInstaller) getCommand(name string) (cmd string) {
 	cmds := make(map[string]string)
 	cmds = map[string]string{
 		initMaster0: `kubeadm init --config=/root/kubeadm-config.yaml --experimental-upload-certs`,
-		initMasters: fmt.Sprintf("kubeadm join %s:6443 --token %s --discovery-token-ca-cert-hash %s --experimental-control-plane --certificate-key %s", s.Masters[0], JoinToken, TokenCaCertHash, CertificateKey),
+		initMasters: fmt.Sprintf("kubeadm join %s:6443 --token %s --discovery-token-ca-cert-hash %s --experimental-control-plane --certificate-key %s", IpFormat(s.Masters[0]), JoinToken, TokenCaCertHash, CertificateKey),
 	}
 
 	if strings.HasPrefix(Version, "v1.15") {

--- a/install/utils.go
+++ b/install/utils.go
@@ -30,6 +30,11 @@ var (
 
 const oneMBByte = 1024 * 1024
 
+func IpFormat(host string) string {
+	ipAndPort := strings.Split(host, ":")
+	return ipAndPort[0]
+}
+
 func AddrReformat(host string) string {
 	if strings.Index(host, ":") == -1 {
 		host = fmt.Sprintf("%s:22", host)

